AC-659 Added contact us screen

rishabh-997 opened the pull request:
<!--- Add a pull request title above in this format -->
<!--- real example: 'AC-204 Applying MVP Model' -->
<!--- 'AC-JiraIssueNumber JiraIssueTitle' -->

## Description of what I changed
1. I added contact us screen with links to email, talks forum, and IRC help in the settings page as well as in the basic menu list.
2. I have added a scroll view to prevent overflow from the screen.
3. I have used **view binding jetpack component** to bind views.
4. Upgraded Gradle build tools to 3.6 and used a new way to store created .apk file

<!--- Describe your changes in detail -->
<!--- It can simply be your commit message, which you must have -->

## Issue I worked on
<!--- This project only accepts pull requests related to open issues -->
<!--- Want a new feature or change? Discuss it in an issue first -->
<!--- Found a bug? Point us to the issue/or create one so we can reproduce it -->
<!--- Just add the issue number at the end: -->
JIRA Issue: https://issues.openmrs.org/browse/AC-659

## Checklist: I completed these to help reviewers :)
<!--- Put an `x` in the box if you did the task -->
<!--- If you forgot a task please follow the instructions below -->
- [x] My pull request only contains **ONE single commit**
(the number above, next to the 'Commits' tab is 1).
<!--- No? -> [read here](https://wiki.openmrs.org/display/docs/Pull+Request+Tips) on how to squash multiple commits into one -->

- [ ] I have **added tests** to cover my changes. (If you refactored
existing code that was well tested you do not have to add tests)
<!--- No? -> write tests and add them to this commit `git add . && git commit --amend`-->

- [x] All new and existing **tests passed**.
<!--- No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works. -->

- [x] My pull request is **based on the latest changes** of the master branch.
<!--- No? Unsure? -> execute command `git pull --rebase upstream master` -->

rishabh-997 commented:
![Screenshot_2020-03-19-02-44-51-02_0b570628ec88a4202b7916ec2df1bebf](https://user-images.githubusercontent.com/45125121/77009505-0ba72f80-698e-11ea-9e16-a7f84724e35d.jpg)
![Screenshot_2020-03-19-02-43-57-78_0b570628ec88a4202b7916ec2df1bebf](https://user-images.githubusercontent.com/45125121/77009511-0e098980-698e-11ea-98f7-34e17b0c2b23.jpg)
![Screenshot_2020-03-19-02-44-07-48_0b570628ec88a4202b7916ec2df1bebf](https://user-images.githubusercontent.com/45125121/77009518-0ea22000-698e-11ea-833e-12b84c6be338.jpg)

rishabh-997 commented:
Howzz that for designs @f4ww4z sir. Have a look at the code.

f4ww4z commented on openmrs-client/src/main/AndroidManifest.xml:
```diff
@@ -268,7 +268,12 @@
+        <activity
+            android:name=".activities.community.contact.ContactUsActivity"
+            android:theme="@style/AppThemeOrig"
+            android:label="Contact Us"/>
```
Move this to `strings.xml` so we can localize it

f4ww4z commented on openmrs-client/src/main/java/org/openmrs/mobile/activities/community/contact/ContactUsActivity.java:
```diff
@@ -0,0 +1,98 @@
+            try {
+                startActivity(i);
+            } catch (android.content.ActivityNotFoundException ex) {
+                ToastUtil.showShortToast(this, ToastUtil.ToastType.ERROR, "There are no email clients installed.");
```
`"There are no email clients installed."` Move this in `strings.xml`.

f4ww4z commented on openmrs-client/src/main/java/org/openmrs/mobile/activities/community/contact/ContactUsActivity.java:
```diff
@@ -0,0 +1,98 @@
+        }
+
+        mailButton.setOnClickListener(v -> {
+            Intent i = new Intent(Intent.ACTION_SENDTO);
```
```suggestion
            Intent sendMailIntent = new Intent(Intent.ACTION_SENDTO);
```

f4ww4z commented on openmrs-client/src/main/java/org/openmrs/mobile/activities/community/contact/ContactUsContract.java:
```diff
@@ -0,0 +1,25 @@
+package org.openmrs.mobile.activities.community.contact;
+
+public class ContactUsContract {
+    interface view {
```
Should be capitalized to match convention. `View`

f4ww4z commented on openmrs-client/src/main/java/org/openmrs/mobile/activities/community/contact/ContactUsContract.java:
```diff
@@ -0,0 +1,25 @@
+
+    }
+
+    interface presenter {
```
See above comment

f4ww4z commented on openmrs-client/src/main/java/org/openmrs/mobile/activities/community/contact/ContactUsPresenter.java:
```diff
@@ -0,0 +1,23 @@
+package org.openmrs.mobile.activities.community.contact;
+
+public class ContactUsPresenter implements ContactUsContract.presenter {
+    ContactUsContract.view mvpview;
```
Rename to e.g. `contactUsView`

f4ww4z commented on openmrs-client/src/main/java/org/openmrs/mobile/activities/settings/SettingsContract.java:
```diff
@@ -36,6 +36,8 @@
 
         void chooseLanguage(String[] languageList);
+
+        void contactUs();
```
```suggestion
        void setUpContactUsButton();
```
should be the name of an action

f4ww4z commented on openmrs-client/src/main/res/values-hi/strings.xml:
```diff
@@ -359,4 +359,14 @@
+	<string name="connect">जुडिये</string>
+	<string name="email_us">ईमेल करे</string>
+	<string name="help_message">हमारी टीम हमेशा किसी भी सवाल का जवाब देने के लिए होती है। हमें एक ईमेल ड्रॉप करें या फ़ोरम पर पहुंचें और हम जल्द से जल्द जवाब देने की पूरी कोशिश करेंगे।</string>
+	<string name="contact_forum_url">https://talk.openmrs.org/</string>
```
Are these fixed strings (like urls and emails) need to be set in every translation file? If not, then don't redefine them, this goes for `helpdesk_mail`, `contact_forum_url` and `contact_irc_url`

f4ww4z commented on openmrs-client/src/main/res/values/strings.xml:
```diff
@@ -360,4 +360,14 @@
+	<string name="helpdesk_mail">helpdesk@openmrs.com</string>
+	<string name="go_to_forum">Go To Forum</string>
+	<string name="go_to_irc">Go To IRC</string>
+	<string name="connect">Connect</string>
```
I think you should change this to 
```suggestion
	<string name="open">Open</string>
```
as it will be clearer to the user. `connect` might be different to what the user is thinking. Let me know your thoughts though.

rishabh-997 commented:
> Great work so far @rishabh-997 , can you use the new [view binding lib](https://developer.android.com/topic/libraries/view-binding#java) instead of butterknife? [See this note from the butterknife team](https://github.com/JakeWharton/butterknife#butter-knife). Note you should only use view binding in this ContactUs screen for now, we can file a new issue to implement it in all screens later.
> 
> Note yo

i was planning to use this one but was not confident as i never used it before. will prepare the PR with it..

rishabh-997 commented:
> Great work so far @rishabh-997 , can you use the new [view binding lib](https://developer.android.com/topic/libraries/view-binding#java) instead of butterknife? [See this note from the butterknife team](https://github.com/JakeWharton/butterknife#butter-knife). Note you should only use view binding in this ContactUs screen for now, we can file a new issue to implement it in all screens later.
> 
> Note yo

Added the view binding feature but to implement it, the following necessary changes had to be done:

1. Upgraded distributionUrl to fetch Gradle services 5.6.4
2. Upgraded Gradle classpath to 3.6.1
3. Updated kotlin classpath 
4. Changed the code of adding generated release apk to release folder as it was throwing following error in 3.6 Gradle version: `groovy.lang.MissingMethodException` ... [reference](https://stackoverflow.com/a/55682025)

f4ww4z commented:
@rishabh-997 I decided to delete `Project.xml` as we're using the eclipse formatter anyway so it's not important. [Follow this code style](https://github.com/openmrs/openmrs-contrib-android-client#code-style) if you haven't already, then fix the merge conflict.

currently testing this Pr thoroughly

rishabh-997 commented:
> @rishabh-997 I decided to delete `Project.xml` as we're using the eclipse formatter anyway so it's not important. [Follow this code style](https://github.com/openmrs/openmrs-contrib-android-client#code-style) if you haven't already, then fix the merge conflict.
> 
> currently testing this Pr thoroughly

resolved the conflict.

f4ww4z commented on openmrs-client/src/main/res/menu/basic_menu.xml:
```diff
@@ -33,6 +33,11 @@
+        android:id="@+id/actionContact"
+        android:orderInCategory="100"
+        android:title="@string/contact_us"
+        app:showAsAction="never" />
```
Show contact us menu item when user is in settings page.

rishabh-997 commented:
> @rishabh-997 The contact us screen option is not showing:
> 
> ![image](https://user-images.githubusercontent.com/21199242/77176236-bf2f3180-6afe-11ea-87e7-b5e277178a6e.png)
> 
> @rishabh-997 also, one last thing - it's better to show this also when user cannot login, so in the login page, when the user clicks 'unable to login' show this contact us screen instead of the pop up message.

i logged out of app to test this feature but after logging out, the app started crashing as discussed here
[talk thread](https://talk.openmrs.org/t/are-there-any-crashes-in-the-latest-android-client-github-build/27457?u=rishabh997)

AryanGanotra07 commented:
> > @rishabh-997 The contact us screen option is not showing:
> > ![image](https://user-images.githubusercontent.com/21199242/77176236-bf2f3180-6afe-11ea-87e7-b5e277178a6e.png)
> > @rishabh-997 also, one last thing - it's better to show this also when user cannot login, so in the login page, when the user clicks 'unable to login' show this contact us screen instead of the pop up message.
> 
> i logged out of app to test this feature but after logging out, the app started crashing as discussed here
> [talk thread](https://talk.openmrs.org/t/are-there-any-crashes-in-the-latest-android-client-github-build/27457?u=rishabh997) so working on resolving that first, so i can test this on login page.
> Here is the link to the ticket [AC-723](https://issues.openmrs.org/browse/AC-723?jql=project%20%3D%20AC%20AND%20resolution%20%3D%20Unresolved%20ORDER%20BY%20updated%20DESC%2C%20priority%20DESC)

@rishabh-997 @f4ww4z  . I have resolved the issue [AC-723](https://issues.openmrs.org/projects/AC/issues/AC-723). Please review the [PR]( https://github.com/openmrs/openmrs-contrib-android-client/pull/716)

AryanGanotra07 commented:
Great design though @rishabh-997 👍 .

f4ww4z commented:
@rishabh-997 I still cannot see the contact us list item here. Check the settings fragment.

![image](https://user-images.githubusercontent.com/21199242/77620651-c9758380-6f75-11ea-8e0b-5de61578e1f6.png)

rishabh-997 commented:
> @rishabh-997 I still cannot see the contact us list item here. Check the settings fragment.
> 
> ![image](https://user-images.githubusercontent.com/21199242/77620651-c9758380-6f75-11ea-8e0b-5de61578e1f6.png)

actually i was stuck on login page, any changes done there arent reflected on app, will see the settings fragment once i am done with this login thing

rishabh-997 commented:
@f4ww4z, there is one problem I am facing for long, any change I do on app code isn't reflected at all in the build, is it because of migration to Gradle 3.6 ??? I tried rebuilding and invalidate caches/restart but nothing seems to work. I am trying to find the issue but the internet seems to have no answer. The app is building for another branch ( lower Gradle ). Or is it because of this [line](https://github.com/openmrs/openmrs-contrib-android-client/pull/710/files#diff-cc5437861bff23b3bb39d227f1f32eaeR76)

rishabh-997 commented:
So finally, the problem was with the updated way of storing the generated app, this branch created the app with name `android-client-debug-version-2.8.1.apk` whereas other branches created with name `android-client-debug.apk` which caused the previous app to install every time rather than the new one, thus not reflecting any change from contact us activity.

You can delete the app from release folder and then rebuild the project, it will show all the changes as I have also updated the code of generating apk
@f4ww4z

f4ww4z commented on openmrs-client/src/main/res/layout/actvity_contact_us.xml:
```diff
@@ -0,0 +1,184 @@
+
+		<ImageView
+			android:id="@+id/contact_image"
+			android:layout_width="200dp"
```
This should not be a fixed value.

rishabh-997 replied:
okay, will convert it to constraint layout

f4ww4z commented on openmrs-client/src/main/res/layout/actvity_contact_us.xml:
```diff
@@ -0,0 +1,184 @@
+
+				<TextView
+					android:id="@+id/contact_email_text"
+					android:layout_width="200dp"
```
This should not be a fixed value. Change to `wrap_content`

f4ww4z commented on openmrs-client/src/main/res/layout/actvity_contact_us.xml:
```diff
@@ -0,0 +1,184 @@
+
+				<TextView
+					android:id="@+id/talks_text"
+					android:layout_width="200dp"
```
See above comment.

f4ww4z commented on openmrs-client/src/main/res/layout/actvity_contact_us.xml:
```diff
@@ -0,0 +1,184 @@
+
+				<TextView
+					android:id="@+id/irc_text"
+					android:layout_width="200dp"
```
See above comment.

rishabh-997 commented:
Migrated activity completely to constraint layout @f4ww4z sr
